prboom2/src/r_defs.h

```diff
@@ -638,4 +662,87 @@ extern fixed_t Sky1ColumnOffset;
 extern fixed_t Sky2ColumnOffset;
 extern dboolean DoubleSky;
 
+
+// A plane which can be a bleed source during rendering.
+// Everything needed to render it in one place.
+typedef struct
+{
+  sector_t dummysector;
+  visplane_t dummyplane;
+  struct component_s* component;
+  sector_t* sector;
+  visplane_t* plane;
+  int lightlevel;
+  int validcount;
+} sourceplane_t;
+
+// A bleed target in a component
+typedef struct
+{
+  struct component_s* component;
+  sourceplane_t* source;
+  int depth;
+} bleedtarget_t;
+
+// A (potential) flat bleed between components
+typedef struct
+{
+  sourceplane_t* source;
+  bleedtarget_t* target;
+  int depth;
+  dboolean ceiling;
+} bleed_t;
+
+// Component flags
+typedef enum compflags_e {
+  COMPF_NONE = 0x0,
+  // Component is degenerate (contains a single degenerate subsector)
+  COMPF_DEGENERATE = 0x1,
+  // Component is eligible for flat "bleed-through"
+  COMPF_BLEED_THROUGH = 0x2,
+  // Component is interior (has no segs on single-sided linedefs)
+  COMPF_INTERIOR = 0x4,
+  // Generic mark for recursive algorithms to avoid cycles or repeating work.
+  // Should always be unset when done.
+  COMPF_MARK = 0x8
+} compflags_t;
```

Comment:
The word "comp" is used throughout the code base for (demo) compatibility so this needs a different name. Component is also too generic, so we need a fancy name for this structure 😄

Reply:
It's technically a [connected component](https://en.wikipedia.org/wiki/Component_\(graph_theory\)) in graph theory, which is where I got the name.  I can think of something else.

Reply:
Any idea for a new name?  Element?  Constituent?